core/kernel/panic.c

```diff
@@ -4,11 +4,76 @@
  * Copyright (c) 2014, STMicroelectronics International N.V.
  */
 
+#include <initcall.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
 #include <kernel/panic.h>
 #include <kernel/thread.h>
 #include <kernel/unwind.h>
 #include <trace.h>
 
+static_assert(!IS_ENABLED(CFG_HALT_CORES_ON_PANIC) ||
+	      (CFG_HALT_CORES_ON_PANIC_SGI >= 8 &&
+	       CFG_HALT_CORES_ON_PANIC_SGI < 16));
+
+#ifdef ARM32
+#define ARM_CNTXCTL_IMASK	BIT(1)
+
+static void mask_timer_interrupts(void)
+{
+	write_cntp_ctl(read_cntp_ctl() | ARM_CNTXCTL_IMASK);
+	write_cntv_ctl(read_cntv_ctl() | ARM_CNTXCTL_IMASK);
+}
+#endif /* ARM32 */
+
+static enum itr_return __noreturn
+multi_core_panic_it_handler(struct itr_handler *hdl  __unused)
+{
+	IMSG("Halting CPU %zu", get_core_pos());
+
+#ifdef ARM32
+	mask_timer_interrupts();
+#endif /* ARM32 */
+
+	while (true)
+		cpu_idle();
+}
+
+static struct itr_handler multi_core_panic_handler = {
+	.it = CFG_HALT_CORES_ON_PANIC_SGI,
+	.handler = multi_core_panic_it_handler,
+};
+DECLARE_KEEP_PAGER(multi_core_panic_handler);
+
+static void notify_other_cores(void)
+{
+	interrupt_raise_sgi(interrupt_get_main_chip(),
+			    CFG_HALT_CORES_ON_PANIC_SGI,
+			    ITR_CPU_MASK_TO_OTHER_CPUS);
+}
+
+static TEE_Result init_multi_core_panic_handler(void)
+{
+	static bool already_registered;
```

Comment:
This isn't needed any longer.

Reply:
Right, final call of primary boot, thanks